In [1]:
import pandas as pd

In [2]:
train = pd.read_csv('/Users/jaemin/Documents/dacon/funda_train.csv')

In [3]:
train.head()

store_id  card_id card_company transacted_date transacted_time  \
0         0        0            b      2016-06-01           13:13   
1         0        1            h      2016-06-01           18:12   
2         0        2            c      2016-06-01           18:52   
3         0        3            a      2016-06-01           20:22   
4         0        4            c      2016-06-02           11:06   

   installment_term region type_of_business       amount  
0                 0    NaN           기타 미용업  1857.142857  
1                 0    NaN           기타 미용업   857.142857  
2                 0    NaN           기타 미용업  2000.000000  
3                 0    NaN           기타 미용업  7857.142857  
4                 0    NaN           기타 미용업  2000.000000

### Q4. 요일 별 매출은 어떻게 변하나요?

In [4]:
pd.to_datetime(train['transacted_date'])  # object->datetime 타입 변경

0         2016-06-01
1         2016-06-01
2         2016-06-01
3         2016-06-01
4         2016-06-02
5         2016-06-02
6         2016-06-02
7         2016-06-02
8         2016-06-02
9         2016-06-02
10        2016-06-02
11        2016-06-03
12        2016-06-03
13        2016-06-03
14        2016-06-04
15        2016-06-04
16        2016-06-04
17        2016-06-04
18        2016-06-04
19        2016-06-04
20        2016-06-04
21        2016-06-05
22        2016-06-05
23        2016-06-05
24        2016-06-06
25        2016-06-06
26        2016-06-06
27        2016-06-06
28        2016-06-09
29        2016-06-09
             ...    
6556583   2019-02-26
6556584   2019-02-26
6556585   2019-02-26
6556586   2019-02-26
6556587   2019-02-26
6556588   2019-02-27
6556589   2019-02-27
6556590   2019-02-27
6556591   2019-02-27
6556592   2019-02-27
6556593   2019-02-27
6556594   2019-02-27
6556595   2019-02-27
6556596   2019-02-27
6556597   2019-02-27
6556598   2019-02-28
6556599   201

In [5]:
pd.to_datetime(train['transacted_date'])[0].year  # 첫번째 값에서 연도 뽑아오기

2016

In [6]:
pd.to_datetime(train['transacted_date'])[50000].month

7

In [7]:
pd.to_datetime(train['transacted_date'])[100000].day

6

In [8]:
pd.to_datetime(train['transacted_date'])[1000001]

Timestamp('2017-02-07 00:00:00')

In [9]:
pd.to_datetime(train['transacted_date'])[1000001].weekday()  # 요일 뽑아오기 (0:월, 1:화, ... 6:일)

1

In [8]:
train['weekday'] = pd.to_datetime(train['transacted_date']).apply(lambda x: x.weekday())  
# apply : 컬럼의 원소 하나하나에 lambda 함수 적용
# lambda : 함수를 간단히 표현하는 방법

In [11]:
train.head(3)

store_id  card_id card_company transacted_date transacted_time  \
0         0        0            b      2016-06-01           13:13   
1         0        1            h      2016-06-01           18:12   
2         0        2            c      2016-06-01           18:52   

   installment_term region type_of_business       amount  weekday  
0                 0    NaN           기타 미용업  1857.142857        2  
1                 0    NaN           기타 미용업   857.142857        2  
2                 0    NaN           기타 미용업  2000.000000        2

In [12]:
day = []
for i in train['weekday']:
    if i==0:
        day.append('월')
    elif i==1:
        day.append('화')
    elif i==2:
        day.append('수')
    elif i==3:
        day.append('목')
    elif i==4:
        day.append('금')
    elif i==5:
        day.append('토')
    else:
        day.append('일')

train['day'] = day

In [13]:
train.head(3)

store_id  card_id card_company transacted_date transacted_time  \
0         0        0            b      2016-06-01           13:13   
1         0        1            h      2016-06-01           18:12   
2         0        2            c      2016-06-01           18:52   

   installment_term region type_of_business       amount  weekday day  
0                 0    NaN           기타 미용업  1857.142857        2   수  
1                 0    NaN           기타 미용업   857.142857        2   수  
2                 0    NaN           기타 미용업  2000.000000        2   수

In [14]:
train.groupby('day')['amount'].sum()

day
금    1.100012e+10
목    1.043264e+10
수    1.041080e+10
월    9.691523e+09
일    6.336814e+09
토    1.028775e+10
화    1.025935e+10
Name: amount, dtype: float64

In [15]:
# 한 컬럼의 값들은 시리즈를 이루고 있음
# 시리즈도 리스트처럼 인덱스로 값을 가져올 수 있음
pd.Series([1,2,3,4])[0]  

1

In [16]:
s = pd.Series([1,2,3,4], index=['A','B','C','D'])  
# s = {'A':1, 'B':2, 'C':3, 'D':4} 라는 딕셔너리와 같은 형태
print(s)

A    1
B    2
C    3
D    4
dtype: int64


In [17]:
s['A']  

1

In [18]:
# reset_index() : 인덱스를 'index'라는 이름을 가진 컬럼의 값들로 만듦
# set_index('column') : 특정 컬럼의 값들을 인덱스로 만듦

##  train에 있고 baseline에는 없는 store_id 찾기 

In [19]:
baseline = train[(train['transacted_date'] >= '2018-03-01') & (train['transacted_date'] <= '2018-05-31')]

In [20]:
train['store_id'].nunique()

1967

In [21]:
baseline['store_id'].nunique()

1955

### 풀이1. set

In [22]:
train_id = set(train['store_id'].unique())  # set 자체가 중복을 허용 안 하기 때문에 unique() 안 써도 됨

In [23]:
baseline_id = set(baseline['store_id'])

In [24]:
train_id - baseline_id

{271, 279, 515, 588, 706, 795, 999, 1379, 1435, 2081, 2094, 2119}

### 풀이2. list와 for문

In [25]:
train_id_list = list(train['store_id'].unique())
baseline_id_list =list(baseline['store_id'].unique())

In [26]:
missing = []
for i in train_id_list:
    if i not in baseline_id_list:       
        missing.append(i)
print(missing)

[271, 279, 515, 588, 706, 795, 999, 1379, 1435, 2081, 2094, 2119]


### 풀이3. isin()

In [27]:
train['store_id'].isin([0,1,2,3]).head()

0    True
1    True
2    True
3    True
4    True
Name: store_id, dtype: bool

In [28]:
train[train['store_id'].isin([0,1,2,3])].head(3) # store_id가 0,1,2,3인 것만 출력

store_id  card_id card_company transacted_date transacted_time  \
0         0        0            b      2016-06-01           13:13   
1         0        1            h      2016-06-01           18:12   
2         0        2            c      2016-06-01           18:52   

   installment_term region type_of_business       amount  weekday day  
0                 0    NaN           기타 미용업  1857.142857        2   수  
1                 0    NaN           기타 미용업   857.142857        2   수  
2                 0    NaN           기타 미용업  2000.000000        2   수

In [29]:
train[~train['store_id'].isin(baseline['store_id'])].head()

store_id  card_id card_company transacted_date transacted_time  \
779947       271   522127            a      2016-06-01           13:05   
779948       271   522128            f      2016-06-01           13:06   
779949       271   522129            g      2016-06-01           13:08   
779950       271   522130            d      2016-06-01           13:15   
779951       271   522131            b      2016-06-01           14:38   

        installment_term region type_of_business       amount  weekday day  
779947                 0  인천 남구              NaN  2285.714286        2   수  
779948                 0  인천 남구              NaN  8414.285714        2   수  
779949                 0  인천 남구              NaN  1350.000000        2   수  
779950                 0  인천 남구              NaN  1428.571429        2   수  
779951                 0  인천 남구              NaN  6285.714286        2   수

In [30]:
missing2 = train[~train['store_id'].isin(baseline['store_id'])]['store_id'].unique()
print(missing2)

[ 271  279  515  588  706  795  999 1379 1435 2081 2094 2119]


# 2019년 3~5월 매출을 예측하는 새로운 아이디어

### 1. baseline에서 빠진 store_id에 해당하는 값 예측 후 baseline과 concat
(2019년 1월부터 2월까지 매출의 합 / 2019년 1~2월의 일 수(59)) * 2019년 3~5월의 일 수(92)

In [31]:
train.head(3)

store_id  card_id card_company transacted_date transacted_time  \
0         0        0            b      2016-06-01           13:13   
1         0        1            h      2016-06-01           18:12   
2         0        2            c      2016-06-01           18:52   

   installment_term region type_of_business       amount  weekday day  
0                 0    NaN           기타 미용업  1857.142857        2   수  
1                 0    NaN           기타 미용업   857.142857        2   수  
2                 0    NaN           기타 미용업  2000.000000        2   수

In [32]:
train_jf = train[(train['transacted_date']>='2019-01-01') & (train['transacted_date']<='2019-02-28')].groupby(['store_id'])['amount'].sum().reset_index()
train_jf.head(3)

store_id        amount
0         0  1.198143e+06
1         1  2.257143e+05
2         2  9.047143e+05

In [33]:
missing

[271, 279, 515, 588, 706, 795, 999, 1379, 1435, 2081, 2094, 2119]

In [34]:
train_jf['forecast'] = (train_jf['amount']/59) * 92
train_jf.head(3)

store_id        amount      forecast
0         0  1.198143e+06  1.868291e+06
1         1  2.257143e+05  3.519613e+05
2         2  9.047143e+05  1.410741e+06

In [76]:
train_jf[train_jf['store_id'].isin(missing)].to_csv('/Users/jaemin/Documents/dacon/0813_2.csv')

store_id        amount      forecast
249        271  1.024661e+07  1.597776e+07
257        279  2.295000e+05  3.578644e+05
474        515  6.595714e+04  1.028484e+05
545        588  3.897143e+04  6.076901e+04
652        706  4.069143e+05  6.345104e+05
734        795  0.000000e+00  0.000000e+00
1263      1379  1.016071e+06  1.584383e+06
1313      1435  7.361429e+05  1.147884e+06
1891      2081  8.412086e+05  1.311715e+06
1903      2094  2.816429e+05  4.391719e+05

In [35]:
train_0812 = pd.read_csv('/Users/jaemin/Documents/dacon/0812_1.csv')
train_0812.head()

store_id        amount
0         0  2.171000e+06
1         1  2.977143e+05
2         2  9.461429e+05
3         4  2.475643e+06
4         5  1.111857e+06

In [36]:
train_0813 = pd.read_csv('/Users/jaemin/Documents/dacon/0813_2.csv')
train_0813_2 = train_0813[['store_id','forecast']]
train_0813_2.head()

store_id      forecast
0       271  1.597776e+07
1       279  3.578644e+05
2       515  1.028484e+05
3       588  6.076901e+04
4       706  6.345104e+05

In [37]:
train_0812.columns = ['store_id','forecast']
train_0812.head()

store_id      forecast
0         0  2.171000e+06
1         1  2.977143e+05
2         2  9.461429e+05
3         4  2.475643e+06
4         5  1.111857e+06

In [38]:
pd.concat([train_0812,train_0813_2])

store_id      forecast
0            0  2.171000e+06
1            1  2.977143e+05
2            2  9.461429e+05
3            4  2.475643e+06
4            5  1.111857e+06
5            6  6.033929e+06
6            7  1.644929e+06
7            8  3.449000e+06
8            9  1.974000e+06
9           10  1.103857e+06
10          11  1.023286e+06
11          12  6.784486e+05
12          13  3.136757e+06
13          14  1.056986e+07
14          15  3.495500e+06
15          16  1.529714e+06
16          17  8.347429e+05
17          18  1.460529e+06
18          19  4.526571e+06
19          20  2.700500e+06
20          22  7.015714e+05
21          23  1.696429e+06
22          24  1.144439e+07
23          25  3.048571e+05
24          26  1.113714e+06
25          27  6.842593e+06
26          28  4.070300e+06
27          29  2.926571e+06
28          30  7.284614e+05
29          31  2.122857e+06
...        ...           ...
1935      2115  3.199986e+06
1936      2116  4.241214e+06
1937      2117  5.175714e+05
1938      2118  2.772571e+06
1939      2120  1.579367e+07
1940      2121  8.320414e+06
1941      2122  8.228896e+06
1942      2123  2.221714e+06
1943      2124  1.496643e+06
1944      2125  7.291286e+06
1945      2126  4.500000e+03
1946      2127  1.296539e+06
1947      2128  1.360000e+06
1948      2129  9.041000e+05
1949      2130  1.724571e+05
1950      2132  2.518143e+06
1951      2133  1.418629e+06
1952      2134  5.340714e+05
1953      2135  1.491000e+06
1954      2136  6.743929e+06
0          271  1.597776e+07
1          279  3.578644e+05
2          515  1.028484e+05
3          588  6.076901e+04
4          706  6.345104e+05
5          795  0.000000e+00
6         1379  1.584383e+06
7         1435  1.147884e+06
8         2081  1.311715e+06
9         2094  4.391719e+05

[1965 rows x 2 columns]

In [141]:
pd.concat([train_0812,train_0813_2]).set_index('store_id').to_csv('0813_3.csv')

In [39]:
added = list(train_0813['store_id'])
added

[271, 279, 515, 588, 706, 795, 1379, 1435, 2081, 2094]

In [40]:
missing

[271, 279, 515, 588, 706, 795, 999, 1379, 1435, 2081, 2094, 2119]

In [41]:
# missing에 있는 id 중 999와 2119에 해당하는 가게는 2019년 1,2월에 매출이 없는 것 같다.
set(missing) - set(added)

{999, 2119}

In [42]:
train.head()

store_id  card_id card_company transacted_date transacted_time  \
0         0        0            b      2016-06-01           13:13   
1         0        1            h      2016-06-01           18:12   
2         0        2            c      2016-06-01           18:52   
3         0        3            a      2016-06-01           20:22   
4         0        4            c      2016-06-02           11:06   

   installment_term region type_of_business       amount  weekday day  
0                 0    NaN           기타 미용업  1857.142857        2   수  
1                 0    NaN           기타 미용업   857.142857        2   수  
2                 0    NaN           기타 미용업  2000.000000        2   수  
3                 0    NaN           기타 미용업  7857.142857        2   수  
4                 0    NaN           기타 미용업  2000.000000        3   목

In [43]:
train.set_index('store_id').loc[999,:]['transacted_date'].min()

'2016-06-01'

In [44]:
train.set_index('store_id').loc[999,:]['transacted_date'].max()

'2018-02-08'

In [45]:
train.set_index('store_id').loc[2119,:]['transacted_date'].min()

'2017-03-23'

In [46]:
train.set_index('store_id').loc[2119,:]['transacted_date'].max()

'2018-02-27'

In [47]:
# 999번은 2016/6/1 부터 2018/2/8 까지, 2119번은 2017/3/23 부터 2018/2/27 까지 매출이 있다.
# 2019년 3월부터 5월까지의 매출 예측이므로 2017년 3월부터 5월까지의 매출을 가져와야겠다. 

In [48]:
t1= train[train['store_id']==999]
t2 =t1[(t1['transacted_date']>='2017-03-01') & (t1['transacted_date']<='2017-05-31')]
t3 = t2['amount'].sum()
print(t3)

6322585.714285715


In [49]:
t4 = train[train['store_id']==2119]
t5 = t4[(t4['transacted_date']>='2017-03-01') & (t4['transacted_date']<='2017-05-31')]
t6 = t5['amount'].sum()
print(t6)

3514000.0


In [50]:
df1 = pd.DataFrame({'store_id': [999, 2119],'amount': [t3,t6]})
df1

store_id        amount
0       999  6.322586e+06
1      2119  3.514000e+06

In [51]:
train2 = pd.read_csv('/Users/jaemin/Documents/dacon/0813_4.csv')
train2.columns = ['store_id', 'amount']
train2.head()

store_id        amount
0         0  2.171000e+06
1         1  2.977143e+05
2         2  9.461429e+05
3         4  2.475643e+06
4         5  1.111857e+06

In [52]:
df2 = pd.concat([df1, train2])
df2 =df2.sort_values(by='store_id')
df2.head()

store_id        amount
0         0  2.171000e+06
1         1  2.977143e+05
2         2  9.461429e+05
3         4  2.475643e+06
4         5  1.111857e+06

In [119]:
df2.set_index('store_id').to_csv('/Users/jaemin/Documents/dacon/0815_1.csv')

In [53]:
len(df2)

1967

### 2. 요일별로 매출 예측 
92일(2019년 3~5월) 동안 월, 화, 수, 목, 금, 토, 일 갯수를 구해서 (미래의 요일 계산) 가지고 있는 데이터의 요일별 매출을 토대로 예측

예시)
1. store_id 271번의 매출(평균): 월 50만원, 화 20만원, 수 60만원,... , 일 100만원
2. 92일 동안 월요일이 14번, 화 32번, 수 13번... , 일 14번이 있었다
3. 50x14, 20x13, 60x13 ... 

In [3]:
# 미래의 요일 알아보기
dates = list(pd.date_range(start='20190301', end='20190531').astype('str'))  # 2019.3.1~2019.5.31 범위의 날짜를 리스트에 담기
                                                                                                # 날짜형 데이터타입은 시분초까지 다 나오므로 문자열 타입으로 바꿔줌

In [4]:
dates_df = pd.DataFrame(dates)
dates_df.columns = ['future']  # 데이터프레임을 만들고 컬럼네임 붙이기
#dates_df.dtypes
dates_df['weekday'] = pd.to_datetime(dates_df['future']).apply(lambda x: x.weekday())  # 요일 컬럼 추가

In [5]:
dates_df.groupby('weekday')['future'].count()

weekday
0    13
1    13
2    13
3    13
4    14
5    13
6    13
Name: future, dtype: int64

In [13]:
train.head()

store_id  card_id card_company transacted_date transacted_time  \
0         0        0            b      2016-06-01           13:13   
1         0        1            h      2016-06-01           18:12   
2         0        2            c      2016-06-01           18:52   
3         0        3            a      2016-06-01           20:22   
4         0        4            c      2016-06-02           11:06   

   installment_term region type_of_business       amount  weekday  
0                 0    NaN           기타 미용업  1857.142857        2  
1                 0    NaN           기타 미용업   857.142857        2  
2                 0    NaN           기타 미용업  2000.000000        2  
3                 0    NaN           기타 미용업  7857.142857        2  
4                 0    NaN           기타 미용업  2000.000000        3

In [14]:
amount_by_day = pd.DataFrame(train.groupby(['store_id','weekday'])['amount'].mean())
amount_by_day

amount
store_id weekday              
0        0         5123.670490
         1         5861.904762
         2         5143.442623
         3         5191.326531
         4         5286.805556
         5         5953.328710
         6         5554.555681
1        0          938.944873
         1          987.021727
         2          998.439221
         3          954.435107
         4          969.598965
2        0        13829.457364
         1        15556.122449
         2        14581.213307
         3        14223.340040
         4        15598.062954
         5        17138.560687
         6        14505.668934
4        0         5216.186855
         1         5454.081633
         2         5211.995249
         3         5282.211982
         4         5222.582174
         5         4905.003909
         6         4890.856593
5        0         5720.767196
         1         5951.732883
         2         5376.773567
         3         5785.235192
...                        ...
2132     5         3548.207319
         6         3580.734406
2133     0         2705.862434
         1         1346.414747
         2         1868.119938
         3         1828.191577
         4         1563.471778
         5         1961.375661
         6         1140.090226
2134     0         3058.470291
         1         2843.778802
         2         3110.961366
         3         2953.296703
         4         2925.410872
         5         3321.199143
         6         3116.615854
2135     0         4776.433271
         1         4863.095238
         2         4661.947662
         3         4795.987431
         4         4549.100369
         5         4567.790444
         6         4723.577236
2136     0         5506.017792
         1         5547.010775
         2         5625.965251
         3         5489.123055
         4         5614.994934
         5         5861.997292
         6         5794.076655

[13574 rows x 1 columns]